### PR TITLE
feat: [CDS-45342]: Add support for certificates in http step

### PR DESCRIPTION
**Pull request by rohitk-h**
## Describe your changes
- Add support for certificates in http step for CDNG
- Added certificate and certificateKey in yaml.

## Checklist
- [x] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/45868)

### CODE REVIEW

File: 970-api-services-beans/src/main/java/io/harness/beans/HttpCertificateNG.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import lombok.Setter;
+
+/**
+ * Http certificate used with encrypted content.
+ */
+@Getter
+@Setter
+@Builder
+public class HttpCertificateNG implements NestedAnnotationResolver, DecryptableEntity {
+  private String accountId;
```
Comment (yogesh-chauhan) [inline 970-api-services-beans/src/main/java/io/harness/beans/HttpCertificateNG.java:26]: do we need accountID here ?

**Review by rohitk-h — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45868)* status: 8 file(s) to review. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/http/HttpTaskNG.java` line 72 at r1](https://reviewable.io/reviews/harness/harness-core/45868#-NRngyFaCtjs7s4V9hte:-NRoZfKz9Gw0eH73ifye:blkwbpf) ([raw file](https://github.com/harness/harness-core/blob/921402b1595be62e65fe0ac4d47dce9be26542f6/930-delegate-tasks/src/main/java/io/harness/delegate/http/HttpTaskNG.java#L72)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can httpTaskParametersNg.getCertificateNG() or httpTaskParametersNg.getCertificateNG().getCertificate() be null ?
</blockquote></details>

fixed, 
___
*[`956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java` line 694 at r1](https://reviewable.io/reviews/harness/harness-core/45868#-NRnh7Oe1O9PDOsZFM8x:-NRoZgr21Y-RlEvefwzF:boac9c8) ([raw file](https://github.com/harness/harness-core/blob/921402b1595be62e65fe0ac4d47dce9be26542f6/956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java#L694)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

Team should be CDC here..
</blockquote></details>

updated. 
___
*[`970-api-services-beans/src/main/java/io/harness/beans/HttpCertificateNG.java` line 26 at r1](https://reviewable.io/reviews/harness/harness-core/45868#-NRkoBLlKMidmqeadKrh-r1-26:-NRoZi5K7qtGPQ-r1NDl:b72jvta) ([raw file](https://github.com/harness/harness-core/blob/921402b1595be62e65fe0ac4d47dce9be26542f6/970-api-services-beans/src/main/java/io/harness/beans/HttpCertificateNG.java#L26)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

do we need accountID here ?
</blockquote></details>

no, I took it from the CG `HttpCertificate` object, will remove. 
___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java` line 127 at r1](https://reviewable.io/reviews/harness/harness-core/45868#-NRkoBLlKMidmqeadKrk-r1-127:-NRoZmpJ6ZdnS8IC6Qgh:b46ma2x) ([raw file](https://github.com/harness/harness-core/blob/921402b1595be62e65fe0ac4d47dce9be26542f6/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java#L127)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

are we using 2 FFs for this
ENABLE_CERT_VALIDATION and CDS_HTTP_STEP_NG_CERTIFICATE ?
</blockquote></details>

CG has this FF so had to add it. @Fernando-Dourado mentioned that this FF is set to true for all accounts in CG. We can verify but I kept the behavior same as CG. 
___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java` line 160 at r1](https://reviewable.io/reviews/harness/harness-core/45868#-NRkoBLlKMidmqeadKrk-r1-160:-NRoZzSd5WyOCQZwTpkH:b-a6s2bp) ([raw file](https://github.com/harness/harness-core/blob/921402b1595be62e65fe0ac4d47dce9be26542f6/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java#L160)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

is it possible to have certificate being present but key being null ?
</blockquote></details>

yes and it is a valid case
___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java` line 162 at r1](https://reviewable.io/reviews/harness/harness-core/45868#-NRkoBLlKMidmqeadKrk-r1-162:-NRo_2w70SsHU5fokF7Y:b6p80q5) ([raw file](https://github.com/harness/harness-core/blob/921402b1595be62e65fe0ac4d47dce9be26542f6/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java#L162)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

redundant line of code 
</blockquote></details>

removed.

**Comment by rohitk-h** [comment]
*[`930-delegate-tasks/src/main/java/io/harness/http/HttpServiceImpl.java` line 319 at r2](https://reviewable.io/reviews/harness/harness-core/45868#-NRp-xXy5Kbc2D45dr8u:-NRp-xXy5Kbc2D45dr8v:b9g508t) ([raw file](https://github.com/harness/harness-core/blob/9fe69a0e0ed89894eb51777c1ba57ae07926110c/930-delegate-tasks/src/main/java/io/harness/http/HttpServiceImpl.java#L319)):*
> ```Java
>     // WHEN INSECURE WE TRUST ALL CERTIFICATE IGNORING THE CA VERIFICATION
>     // To support self-signed certificates set the trustStrategy to TrustSelfSignedStrategy
>     final TrustStrategy trustStrategy = insecure ? (x509Certificates, s) -> true : new TrustSelfSignedStrategy();
> ```

@Fernando-Dourado I had to had this so as to test locally. Also, this will allow support for self-signed certificates for customers.

File: 930-delegate-tasks/src/main/java/io/harness/delegate/http/HttpTaskNG.java [hunk 2]
```diff
@@ -55,6 +56,19 @@ public HttpStepResponse run(TaskParameters parameters) throws IOException {
...
+    if (httpTaskParametersNg.getCertificateNG() != null) {
+      if (httpTaskParametersNg.getCertificateNG().getCertificate() != null) {
+        httpCertificate.cert(httpTaskParametersNg.getCertificateNG().getCertificate().toCharArray());
+      }
+      if (httpTaskParametersNg.getCertificateNG().getCertificateKey() != null) {
+        httpCertificate.certKey(httpTaskParametersNg.getCertificateNG().getCertificateKey().toCharArray());
+      }
+    }
+
+    httpCertificate.keyStoreType(httpTaskParametersNg.getCertificateNG().getKeyStoreType());
```
Comment (yogesh-chauhan) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/http/HttpTaskNG.java:70]: the above logic can be extracted out into a method for better readability

**Review by Fernando-Dourado — DISMISSED** [review]
> Reviewed 2 of 8 files at r1, 3 of 5 files at r2, 4 of 8 files at r3, 3 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45868)* status: 16 file(s) to review. 4. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/io/harness/http/HttpServiceImpl.java` line 319 at r2](https://reviewable.io/reviews/harness/harness-core/45868#-NRp-xXy5Kbc2D45dr8u:-NRrIRPiBuQxcws4wv0d:b-rav7ft) ([raw file](https://github.com/harness/harness-core/blob/9fe69a0e0ed89894eb51777c1ba57ae07926110c/930-delegate-tasks/src/main/java/io/harness/http/HttpServiceImpl.java#L319)):*
<details><summary><i>Previously, rohitk-h (Rohit Karelia) wrote…</i></summary><blockquote>

@Fernando-Dourado I had to had this so as to test locally. Also, this will allow support for self-signed certificates for customers. 
</blockquote></details>

Good solution.
___
*[`970-api-services-beans/src/main/java/io/harness/beans/HttpCertificateNG.java` line 25 at r4](https://reviewable.io/reviews/harness/harness-core/45868#-NRrI3z11ItdjcjBGpTL:-NRrI3z11ItdjcjBGpTM:beoj4u8) ([raw file](https://github.com/harness/harness-core/blob/df2560868907904e88192e41d1c647691a490cf2/970-api-services-beans/src/main/java/io/harness/beans/HttpCertificateNG.java#L25)):*
> ```Java
> @Setter
> @Builder
> public class HttpCertificateNG implements NestedAnnotationResolver, DecryptableEntity {
> ```

It's not possible to use the same class as CG?
___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java` line 127 at r1](https://reviewable.io/reviews/harness/harness-core/45868#-NRkoBLlKMidmqeadKrk-r1-127:-NRrKG_Q2z_ezJAzsx3e:b-2lldvp) ([raw file](https://github.com/harness/harness-core/blob/921402b1595be62e65fe0ac4d47dce9be26542f6/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java#L127)):*
<details><summary><i>Previously, rohitk-h (Rohit Karelia) wrote…</i></summary><blockquote>

CG has this FF so had to add it. @Fernando-Dourado mentioned that this FF is set to true for all accounts in CG. We can verify but I kept the behavior same as CG. 
</blockquote></details>

Is better to keep the same behavior. However, I made a mistake about the PROD values for this FF. We have this FF enabled only for three accounts, two in PROD2 and one in PROD3, to all the other customers we are returning it as FALSE.

**Review by rohitk-h — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45868)* status: 16 file(s) to review. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`970-api-services-beans/src/main/java/io/harness/beans/HttpCertificateNG.java` line 25 at r4](https://reviewable.io/reviews/harness/harness-core/45868#-NRrI3z11ItdjcjBGpTL:-NRsfob1EgNI6U_ztdFE:bj5ylrp) ([raw file](https://github.com/harness/harness-core/blob/df2560868907904e88192e41d1c647691a490cf2/970-api-services-beans/src/main/java/io/harness/beans/HttpCertificateNG.java#L25)):*
<details><summary><i>Previously, Fernando-Dourado (Fernando Dourado) wrote…</i></summary><blockquote>

It's not possible to use the same class as CG?
</blockquote></details>

It is not because secret decryption is different in NG.
___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java` line 127 at r1](https://reviewable.io/reviews/harness/harness-core/45868#-NRkoBLlKMidmqeadKrk-r1-127:-NRsfvKA5ku7TE6p149T:b-oiy11b) ([raw file](https://github.com/harness/harness-core/blob/921402b1595be62e65fe0ac4d47dce9be26542f6/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java#L127)):*
<details><summary><i>Previously, Fernando-Dourado (Fernando Dourado) wrote…</i></summary><blockquote>

Is better to keep the same behavior. However, I made a mistake about the PROD values for this FF. We have this FF enabled only for three accounts, two in PROD2 and one in PROD3, to all the other customers we are returning it as FALSE.
</blockquote></details>

ok, I will make a note in JIRA, these both FF must be enabled for the feature.

**Review by rohitk-h — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45868)* status: 16 file(s) to review. 6. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/http/HttpTaskNG.java` line 70 at r4](https://reviewable.io/reviews/harness/harness-core/45868#-NRkoBLlKMidmqeadKrd-r4-70:-NRt28fEB2aWF95Ds1Ls:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/df2560868907904e88192e41d1c647691a490cf2/930-delegate-tasks/src/main/java/io/harness/delegate/http/HttpTaskNG.java#L70)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

the above logic can be extracted out into a method for better readability
</blockquote></details>

Done.

**Review by hingerabhinav — APPROVED** [review]
> Reviewed 3 of 8 files at r1, 2 of 5 files at r2, 5 of 8 files at r3, 2 of 4 files at r4, 1 of 1 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45868)* status: 6 file(s) to review. 6. unresolved discussions. 1 of 2 approvals obtained.

**Comment by yogesh-chauhan** [comment]
*[`930-delegate-tasks/src/main/java/io/harness/delegate/http/HttpTaskNG.java` line 103 at r7](https://reviewable.io/reviews/harness/harness-core/45868#-NS5T_E7546fadoI-Gk0:-NS5T_E7546fadoI-Gk1:b81nq9p) ([raw file](https://github.com/harness/harness-core/blob/d49bb2bad1d45308306a902883991cba1f9867f6/930-delegate-tasks/src/main/java/io/harness/delegate/http/HttpTaskNG.java#L103)):*
> ```Java
>     }
> 
>     httpCertificate.keyStoreType(httpTaskParametersNg.getCertificateNG().getKeyStoreType());
> ```

this should move into the above if block

**Comment by yogesh-chauhan** [comment]
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java` line 160 at r1](https://reviewable.io/reviews/harness/harness-core/45868#-NRkoBLlKMidmqeadKrk-r1-160:-NS5UGgl1rGEXeQeiJZp:bd0tw6e) ([raw file](https://github.com/harness/harness-core/blob/921402b1595be62e65fe0ac4d47dce9be26542f6/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java#L160)):*
<details><summary><i>Previously, rohitk-h (Rohit Karelia) wrote…</i></summary><blockquote>

yes and it is a valid case
</blockquote></details>

in that case will the below code not throw NPE since the key is null ?
```
    return Optional.of(HttpCertificateNG.builder()
                           .certificate(httpStepParameters.getCertificate().getValue())
                           .certificateKey(httpStepParameters.getCertificateKey().getValue())
                           .build());
```

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 2 of 8 files at r1, 2 of 5 files at r2, 5 of 8 files at r3, 1 of 4 files at r4, 3 of 3 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45868)* status: 2 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained.

**Review by rohitk-h — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45868)* status: 2 file(s) to review. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/http/HttpTaskNG.java` line 103 at r7](https://reviewable.io/reviews/harness/harness-core/45868#-NS5T_E7546fadoI-Gk0:-NS885HE1FABQcS2qvlw:b-dw8hqc) ([raw file](https://github.com/harness/harness-core/blob/d49bb2bad1d45308306a902883991cba1f9867f6/930-delegate-tasks/src/main/java/io/harness/delegate/http/HttpTaskNG.java#L103)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

this should move into the above if block
</blockquote></details>

good catch, updated. 
___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java` line 160 at r1](https://reviewable.io/reviews/harness/harness-core/45868#-NRkoBLlKMidmqeadKrk-r1-160:-NS87Tts0uYL3ENTKAhc:b-x8wzok) ([raw file](https://github.com/harness/harness-core/blob/921402b1595be62e65fe0ac4d47dce9be26542f6/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java#L160)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

in that case will the below code not throw NPE since the key is null ?
```
    return Optional.of(HttpCertificateNG.builder()
                           .certificate(httpStepParameters.getCertificate().getValue())
                           .certificateKey(httpStepParameters.getCertificateKey().getValue())
                           .build());
```
</blockquote></details>

It doesn't because value is wrapped within `ParameterField` object . Tested the use case.

**Review by yogesh-chauhan — DISMISSED** [review]
> Reviewed 1 of 4 files at r4, 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45868)* status: 4 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by Fernando-Dourado — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45868)* status: 4 file(s) to review. 1 of 2 approvals obtained.

**Review by Fernando-Dourado — APPROVED** [review]
> Reviewed 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45868)* status: 4 file(s) to review. 1 of 2 approvals obtained.

**Review by Fernando-Dourado — APPROVED** [review]
> Reviewed 1 of 8 files at r1, 1 of 8 files at r3, 1 of 4 files at r4, 1 of 3 files at r6, 1 of 1 files at r7, 2 of 3 files at r8.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45868)* status: 1 file(s) to review. 1 of 2 approvals obtained.

**Review by Fernando-Dourado — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45868#-:-NSGW4u41NImj2xS0ETk:bnfp4nl)* status: 1 file(s) to review. 1 of 2 approvals obtained.

**Review by Fernando-Dourado — APPROVED** [review]
> Reviewed 3 of 6 files at r10, 3 of 3 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45868)* status: 6 file(s) to review. 1 of 2 approvals obtained.

**Review by sribalij — APPROVED** [review]
> Reviewed 3 of 8 files at r1, 2 of 5 files at r2, 3 of 8 files at r3, 2 of 4 files at r4, 1 of 3 files at r8, 3 of 6 files at r10, 3 of 3 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45868)* status: :shipit: complete! ✓

**Pull request merged**